src/KGBotka/Calc.hs

```diff
@@ -0,0 +1,210 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+module KGBotka.Calc where
+
+import Control.Applicative (Alternative(..))
+import Control.Monad (void)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Except (ExceptT(..), throwE)
+import Data.Char (isAlpha, isDigit)
+import qualified Data.Map.Strict as M
+import qualified Data.Text as T
+import KGBotka.Parser
+import System.Random (randomIO)
+
+data Operator
+  = Add
+  | Sub
+  | Mul
+  | Div
+  | Mod
+  | Pow
+  deriving (Show, Eq)
+
+charToOperator :: Char -> Operator
+charToOperator '+' = Add
+charToOperator '-' = Sub
+charToOperator '*' = Mul
+charToOperator '/' = Div
+charToOperator '%' = Mod
+charToOperator '^' = Pow
+charToOperator _ =
+    -- TODO: Make charToOperator not throw an exception
+    -- For now it is an unsafe function because it indicates
+    -- that operators were added but the parsers were not
+    -- corrected.
+  error
+    "charToOperator: No pattern for operator conversion. \
+    \Maybe you added a new operator and forgot to add a case to the conversion function."
+
+data CalcExpression
+  = BinaryExpression Operator CalcExpression CalcExpression
+  | NegativeExpression CalcExpression
+  | FunctionApplication T.Text [CalcExpression]
+  | ValueExpression Double
+  deriving (Show)
+
+newtype CalcEvalError =
+  CalcEvalError T.Text
+  deriving (Show)
+
+type CalcEval = ExceptT CalcEvalError IO
+
+parseNumber :: Parser Double
+parseNumber = parseFloating <|> parseInteger
+  where
+    parseNumeric :: Parser T.Text
+    parseNumeric = notNull "Expected a numeric value" $ takeWhileP isDigit
+    parseInteger :: Parser Double
+    parseInteger = read . T.unpack <$> parseNumeric
+    -- TODO: parseFloating does not support exponential number format
+    parseFloating :: Parser Double
+    parseFloating = do
+      integerPart <- parseNumeric
+      void $ charP '.'
+      fractionalPart <- parseNumeric
+      return $ read $ T.unpack $ integerPart <> "." <> fractionalPart
+
+parseLine :: Parser CalcExpression
+parseLine = parseExpression <* eof
+
+parseExpression :: Parser CalcExpression
+parseExpression = parseAdditive
+
+parseAdditive :: Parser CalcExpression
+parseAdditive = parseAdditive' <|> parseMultiplicative
+  where
+    parseAdditive' = do
+      left <- parseMultiplicative
+      operator <- charToOperator <$> (charP '+' <|> charP '-')
```

Comment:
@rexim I just pondered about this for a bit. Maybe this could be something like:
`Add <$ charP '+'` 
That way we would not trigger the exception discussed above.

Reply:
Resolved with 
a4ee58a